addons/utm/models/utm.py

```diff
@@ -42,14 +49,57 @@ def _group_expand_stage_ids(self, stages, domain, order):
         stage_ids = stages._search([], order=order, access_rights_uid=SUPERUSER_ID)
         return stages.browse(stage_ids)
 
+    def _merge_utm_campaigns(self, keep_campaign, merge_values={}):
+        if len(self.ids) <= 1:
+            raise UserError(_('Please select more than one campaign from the list.'))
+        if not all(campaign.active for campaign in self):
+            raise UserError(_('Only active campaigns can be merged.'))
+        if not keep_campaign or keep_campaign not in self:
+            raise UserError(_('Kept campaign should be one of the merged campaign.'))
+
+        merge_values['is_website'] = any(campaign.is_website for campaign in self)
+        keep_campaign.write(merge_values)
+
+        # deactive uneeded campaigns, set reference to the merged campaign, 
+        # and update reference of other campaigns if the reference is deactived
+        deactived_campaigns = self - keep_campaign
+        deactived_campaigns.write({'active': False, 'reference_utm_campaign_id': keep_campaign.id})
+        self.search([('reference_utm_campaign_id', 'in', deactived_campaigns.ids)]).write({
+            'reference_utm_campaign_id': keep_campaign.id
+        })
+
+        self._clean_merged_campaigns_references(keep_campaign)
+
+    def _clean_merged_campaigns_references(self, merged_campaign):
+        """After a campaign merge, other modules with m2o linked to deactived campaign should be redirect 
+        to the merged one. Search all fields that is m2o and linked to utm.campaign and not ralated field. 
+        Also igonre fields belong to AbstractModel and models we intentionally ignored."""
+        for field in self.env['ir.model.fields'].search([('ttype', '=', 'many2one'), 
+            ('relation', '=', 'utm.campaign'), ('related', '=', False)]):
+            if not field.model_id.transient:
+                model = self.env[field.model]
+                if model._auto and model._name not in self._get_modules_to_ignore():
+                    self._clean_merged_campaigns_reference(model._name, field.name, merged_campaign)
+
+    def _clean_merged_campaigns_reference(self, model_name, field_name, merged_campaign):
+        model = self.env[model_name]
+        record_to_redirect = model.search([(field_name, 'in', (self - merged_campaign).ids)])
+        for record in record_to_redirect:
+            record.sudo().write({field_name: merged_campaign})
+
+    def _get_modules_to_ignore(self):
```

Comment:
It's more "models" than "modules".

I would go for `_get_ignored_merge_models`